aws/resource_aws_qldb_ledger_test.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"log"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+func init() {
+	resource.AddTestSweepers("aws_qldb_ledger", &resource.Sweeper{
+		Name: "aws_qldb_ledger",
+		F:    testSweepQLDBLedgers,
+	})
+}
+
+func testSweepQLDBLedgers(region string) error {
+	client, err := sharedClientForRegion(region)
+
+	if err != nil {
+		return fmt.Errorf("error getting client: %s", err)
+	}
+
+	conn := client.(*AWSClient).qldbconn
+	req := &qldb.DescribeLedgerInput{}
+	resp, err := conn.DescribeLedger(req)
+	if err != nil {
+		if testSweepSkipSweepError(err) {
+			log.Printf("[WARN] Skipping QLDB Ledger sweep for %s: %s", region, err)
+			return nil
+		}
+		return fmt.Errorf("Error describing QLDB Ledgers: %s", err)
+	}
+
+	if len(aws.StringValue(resp.Name)) == 0 {
+		log.Print("[DEBUG] No aws QLDB Kedgers to sweep")
+		return nil
+	}
+
+	return nil
+}
+
+func TestAccAWSQLDBLedger_basic(t *testing.T) {
+	var qldbCluster qldb.DescribeLedgerOutput
+	rInt := acctest.RandInt()
+	resourceName := "aws_qldb_ledger.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSQLDBLedgerDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSQLDBLedgerConfig(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSQLDBLedgerExists(resourceName, &qldbCluster),
+					resource.TestMatchResourceAttr(resourceName, "arn", regexp.MustCompile(`^arn:[^:]+:qldb:[^:]+:\d{12}:ledger/.+`)),
+					resource.TestMatchResourceAttr(resourceName, "name", regexp.MustCompile("test-ledger-[0-9]+")),
+					resource.TestCheckResourceAttr(resourceName, "permissions_mode", "ALLOW_ALL"),
+					resource.TestCheckResourceAttr(resourceName, "deletion_protection", "false"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"apply_immediately",
+				},
+			},
+		},
+	})
+}
+
+func testAccCheckAWSQLDBLedgerDestroy(s *terraform.State) error {
+	err := testAccCheckAWSClusterDestroyWithProvider(s, testAccProvider)
+	if err != nil {
+		qldberr := err.(awserr.Error)
+		if qldberr.Code() == "ResourceInUseException" {
+			log.Println("Ledger is in creating state. Retrying in 10 seconds...")
+			time.Sleep(time.Second * 10)
+			return testAccCheckAWSQLDBLedgerDestroy(s)
+		}
+	}
+
+	return err
+}
+
+func testAccCheckAWSQLDBLedgerExists(n string, v *qldb.DescribeLedgerOutput) resource.TestCheckFunc {
+	return testAccCheckAWSQLDBLedgerExistsWithProvider(n, v, func() *schema.Provider { return testAccProvider })
```

Comment:
We do not need the indirection here of another function -- the `testAccCheckAWSQLDBLedgerExistsWithProvider` logic can be implemented inside `testAccCheckAWSQLDBLedgerExists` instead (replacing `provider` with `testAccProvider`).